tests/client.rs

```diff
@@ -0,0 +1,88 @@
+use self::support::*;
+use anyhow::{bail, Context, Result};
+use std::fs;
+use warg_client::{
+    storage::{ContentStorage, PublishEntry, PublishInfo, RegistryStorage},
+    Config, FileSystemClient, StorageLockResult,
+};
+
+pub mod support;
+
+fn create_client(config: &Config) -> Result<FileSystemClient> {
+    match FileSystemClient::try_new_with_config(None, config)? {
+        StorageLockResult::Acquired(client) => Ok(client),
+        _ => bail!("failed to acquire storage lock"),
+    }
+}
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 1)]
+async fn client_incrementally_fetches() -> Result<()> {
+    const PACKAGE_COUNT: usize = 2000;
+
+    let (_server, config) = spawn_server(&root().await?, None).await?;
+    let client = create_client(&config)?;
+    let signing_key = support::test_signing_key().parse().unwrap();
+
+    // Store a single component that will be used for every release
+    let bytes =
+        wat::parse_str("(component)").context("failed to parse component for publishing")?;
+    let digest = client
+        .content()
+        .store_content(
+            Box::pin(futures::stream::once(async move { Ok(bytes.into()) })),
+            None,
+        )
+        .await?;
+
+    // Initialize a new package log with releases of the same content
+    // As the number of packages exceeds the limit for fetching in a single request,
+    // this will guarantee that the later call to `upsert` fetches the package log incrementally
+    let mut entries = Vec::with_capacity(PACKAGE_COUNT + 1);
+    entries.push(PublishEntry::Init);
+
+    for i in 1..=PACKAGE_COUNT {
+        entries.push(PublishEntry::Release {
```

Comment:
Making a large number of entries in one update still only makes one record. Does this test what we're trying to test?

Reply:
erm whoops! In my attempt to cut down on the calls on publish, I completely invalidate the premise of the test 🤦 .

Let me also run it before the fix commit to ensure it is actually catching the bug.